### PR TITLE
feat: implement a course registry system

### PR DESCRIPTION
**Pull request by khayss**
## Description
This PR implements the `CourseRegistry` system for the `StudentRegistry` project. The implementation covers the course registry, allows for the creation of a course and each course has a unique id, name, max capacity (maximum number of students that can register). Students can enroll in one or more courses. The course registry is integrated with the student registry.

## Requirements

- Implement a course registration system.
- Expand the `StudentRegistry` struct to include `CourseRegistry` - where students can register for courses. 
- Each course has a unique ID, a name, and a capacity (maximum number of students that can enroll). 
- Students should be able to enroll in multiple courses, and a course should only accept a limited number of students.
- Link the `CourseRegistry` with the `StudentRegistry`
## Changes Made
- Update `StudentRegistry` struct to contain `course_registry`.
```rust
pub struct StudentRegistry {
    pub total_students: Vec<Student>,
    pub course_registry: CourseRegistry,
}
```
- Add `CourseRegistry` struct in `src/student_registry_project/types/basic_types.rs`. This struct contains `courses` map using a BtreeMap and `total_courses`.
```rust
pub struct CourseRegistry {
    pub courses: BTreeMap<u32, Course>,
    pub total_courses: u32,
}
```
- Add `Course` struct in `src/student_registry_project/types/basic_types.rs`. This struct holds `course_name`, `students_ids`, `max_capacity`, and `id` .
```rust
pub struct Course {
    pub name: String,
    pub student_ids: Vec<u32>,
    pub max_capacity: u32,
    pub id: u32,
}
```
- Create `course_registry.rs` file in `src/student_registry_project/implementations/`. This file contains implementation of the `CourseRegistry` and `Course` struct.
```rust
// CourseRegistry implementation
impl CourseRegistry {
    // static method to initializea a new CourseRegistry
    pub fn new() -> CourseRegistry {
        CourseRegistry {
            courses: BTreeMap::new(),
            total_courses: 0,
        }
    }

    // method to add a course to the registry
    pub fn add_course(&mut self, name: String, max_capacity: u32) -> u32 {
        let course_id = self.total_courses + 1;

        let course = Course::new(course_id, name, max_capacity);

        self.total_courses += 1;
        self.courses.insert(course_id, course);

        course_id
    }

    // method to get a course by id
    pub fn get_course_by_id(&mut self, id: u32) -> Option<&mut Course> {
        self.courses.get_mut(&id)
    }
}

// Course implementation
impl Course {
    // static method to initialize a new course
    pub fn new(id: u32, name: String, max_capacity: u32) -> Course {
        Course {
            id,
            name,
            student_ids: Vec::new(),
            max_capacity,
        }
    }

    // method to add a student to a course
    pub fn add_student(&mut self, student_id: u32) -> Result<(), String> {
        // checks if the course has reached max capacity, if it has, returns an error
        if self.student_ids.len() >= self.max_capacity as usize {
            return Err(convert_to_string("Course is full"));
        }
        // checks if the student is already registered for the course
        let student = self.student_ids.iter().find(|&s| s == &student_id);

        match student {
            // if student is already registered, returns an error
            Some(_) => Err(convert_to_string("Student already registered")),

            // if student is not registered, adds the student to the course
            None => {
                self.student_ids.push(student_id);
                Ok(())
            }
        }
    }
}
```
- Add declaration for the `course_registry` module in `src/student_registry_project/implementations/mod.rs`
```rust
`src/student_registry_project/implementations/`
```
- Update `src/main.rs` with code to confirm that the `CourseRegistry` implementation works as expected.
```rust
  // Student Registry
    let mut rust_cohort = StudentRegistry::new_session();
    let _stephanie = rust_cohort.register(
        convert_to_string("Stephanie"),
        convert_to_string("Nwankwo"),
        20,
        5.8,
        Sex::Female,
    );

    // registers a second student to the cohort
    let _rico = rust_cohort.register(
        convert_to_string("rico"),
        convert_to_string("blockheader"),
        22,
        6.1,
        Sex::Male,
    );
    // println!("stephanie as student: {:#?} ", stephanie);
    // println!("stephanie as student: {stephanie:#?} ");

    // student with id 0 is stephanie
    let st_1 = rust_cohort.get_student_by_id(0);
    println!("student with id 0 is: {st_1:#?} ");

    // student with id 1 is rico
    let st_2 = rust_cohort.get_student_by_id(1);
    println!("student with id 1 is: {st_2:#?} ");
    print!("\n\n");

    // create course 1 with max capacity of 1 student
    let course_1 = rust_cohort.add_course(convert_to_string("Introduction To Rust"), 1);

    // create course 2 with max capacity of 3 students
    let course_2 = rust_cohort.add_course(convert_to_string("Advance Cairo"), 3);

    println!("course 1 is created with id: {course_1:#?}");
    println!("course 2 is created with id: {course_2:#?}");
    print!("\n\n");

    // attempt to register stephanie for course 1
    // stephanie student id is 0
    let course_1_reg_result_1 = rust_cohort.register_student_for_course(course_1, 0);

    // attempt to register rico for course 1
    // rico student id is 0
    // registration should return an error because course 1 has max capacity of 1 student
    let course_1_reg_result_2 = rust_cohort.register_student_for_course(course_1, 1);

    println!("stephanie registration result for course_1: {course_1_reg_result_1:#?}");
    println!("rico registration result for course_2: {course_1_reg_result_2:#?}");
    print!("\n\n");

    // attempt to register stephanie for course 1
    // stephanie student id is 0
    // also test that student can register for more than 1 course
    let course_2_reg_result_1 = rust_cohort.register_student_for_course(course_2, 0);

    // attempt to register stephanie for course 1 twice
    // stephanie student id is 0
    // should fail as double registration are not allowed
    let course_2_reg_result_2 = rust_cohort.register_student_for_course(course_2, 0);

    // attempt to register rico for course 1
    // rico student id is 0
    // should return an Ok result since max capacity is not reached
    let course_2_reg_result_3 = rust_cohort.register_student_for_course(course_2, 1);

    println!("stephanie registration result for course_2: {course_2_reg_result_1:#?}");
    println!("stephanie second registration result for course_2: {course_2_reg_result_2:#?}");
    println!("rico registration result for course_2: {course_2_reg_result_3:#?}");
```

## Rationale
- The `CourseRegistry` struct contains two fields - `courses` and `total_courses`. The `courses` field is of type `BTreeMap<u32, Courses>, this allows us to map created courses to a unique id and also for easy retrieval. The `total_courses` field is of type `u32`, this allows us to keep track of current courses.
- The `Course` struct contains four fields - `name`, `student_ids`, `max_capacity`, and `id`. The `name` field is of type `String` and stores the name of the course. The`student_ids` fields is of type `Vec<u32>`, this field stores the ids of students currently enrolled in the course. The `max_capacity` field is of type `u32` and represents the max number of students that can enroll in the course. The `id` field is of type `u32`, it uniquely identifies each course.
## Testing

- [x] carried out tests to confirm that a student can register for a course.
- [x] carried out tests to confirm that a student can resgiter for more than one course.
- [x] carried out tests to confirm student can't register for same course twice.
- [x] carried out tests to confirm student can't register for a course once the max capacity is reached.
- [x] carried out tests to confirm multiple students can register for a course

## Linked Issues
[Assignment 1 (10-08-24)](https://github.com/BlockheaderWeb3-Community/rust-workshop/blob/main/src/student_registry_project/assignment-1.md)

Checklist:

- [x] Code compiles without errors
- [x] Tests have been added/updated to cover changes
- [x] Documentation has been updated (if applicable)
- [x] Code follows the project's coding standards

Kindly review the changes and provide feedback or approval. Thank you!

### CODE REVIEW

File: src/student_registry_project/implementations/course_registry.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
+use std::collections::BTreeMap;
```
Comment (Abeeujah) [inline src/student_registry_project/implementations/course_registry.rs]: Prefer HashMap instead

File: src/student_registry_project/implementations/course_registry.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    // method to add a student to a course
+    pub fn add_student(&mut self, student_id: u32) -> Result<(), String> {
+        // checks if the course has reached max capacity, if it has, returns an error
+        if self.student_ids.len() >= self.max_capacity as usize {
+            return Err(convert_to_string("Course is full"));
+        }
+        // checks if the student is already registered for the course
+        let student = self.student_ids.iter().find(|&s| s == &student_id);
```
Comment (Abeeujah) [inline src/student_registry_project/implementations/course_registry.rs]: Dereference Value
*s == student_id

File: src/student_registry_project/implementations/student_registry.rs [hunk 3]
```diff
@@ -37,4 +41,39 @@ impl StudentRegistry {
...
+    // util function to register student for a course
+    pub fn register_student_for_course(
+        &mut self,
+        course_id: u32,
+        student_id: u32,
+    ) -> Result<(), String> {
+        // checks if the student with the id exists, this way, can't add students that don't exist
+        let student = self.get_student_by_id(student_id);
+
+        match student {
```
Comment (Abeeujah) [inline src/student_registry_project/implementations/student_registry.rs:54]: Confirm the course exists
Confirm the student exists
Enroll the student

File: src/student_registry_project/implementations/course_registry.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// Course implementation
+impl Course {
+    // static method to initialize a new course
+    pub fn new(id: u32, name: String, max_capacity: u32) -> Course {
+        Course {
+            id,
+            name,
+            student_ids: Vec::new(),
```
Comment (Abeeujah) [inline src/student_registry_project/implementations/course_registry.rs]: Vec::with_capacity()

**Review by Abeeujah — CHANGES_REQUESTED** [review]
Great Job, Please Implement the prescribed changes

File: src/student_registry_project/implementations/course_registry.rs (hunk 1 above)
Comment (khayss) [reply src/student_registry_project/implementations/course_registry.rs]: implemented this in the update

File: src/student_registry_project/implementations/course_registry.rs (hunk 2 above)
Comment (khayss) [reply src/student_registry_project/implementations/course_registry.rs]: implemented this in the update.

File: src/student_registry_project/implementations/student_registry.rs (hunk 3 above)
Comment (khayss) [reply src/student_registry_project/implementations/student_registry.rs:54]: Could not implement this due to Rust's ownership rule. Checking the order violates Rust's ownership rule as `course` will first take a mutable reference to self and as such no other reference is allowed till `course` is out of scope.

File: src/student_registry_project/implementations/course_registry.rs (hunk 4 above)
Comment (khayss) [reply src/student_registry_project/implementations/course_registry.rs]: Implemented this in the update

**Pull request merged**